packages/generator/src/options-parser.ts

```diff
@@ -0,0 +1,174 @@
+import { createLogger } from '@sap-cloud-sdk/util';
+import { Options } from 'yargs';
+import { ParsedOptions } from './generator-options';
+const logger = createLogger('generator-options');
+
+type Option = Options & {
+  replacedBy?: string;
+  describe: string;
+};
+
+/**
+ * @internal
+ * Remove defaults from CLI options. This is necessary to handle default setting on our own.
+ * @param options - CLI options, record mapping option name to option config
+ * @returns CLI options without default values
+ */
+export function getOptionsWithoutDefaults<
+  CliOptionsT extends Record<string, Option>
+>(options: CliOptionsT): CliOptionsT {
+  return Object.entries(options).reduce((opts, [name, option]) => {
+    const { default: def, ...optionWithoutDefault } = option;
+    const describe =
+      def === undefined
+        ? optionWithoutDefault.describe
+        : `${optionWithoutDefault.describe}\n[Default: ${def}]`;
+    return {
+      ...opts,
+      [name]: {
+        ...optionWithoutDefault,
+        describe
+      }
+    };
+  }, {} as CliOptionsT);
+}
+
+/**
+ * @internal
+ * Parse options for programmatic and CLI use.
+ * Warn, if deprecated options are used or duplicate.
+ * Adds defaults if unset.
+ * @param options - Available CLI options along with their configuration.
+ * @param userOptions - Options as set by user.
+ * @returns Parsed options with default values.
+ */
+export function parseOptionsWithDefaults<
+  CliOptionsT extends Record<string, Option>,
+  GeneratorOptionsT extends Record<string, any>
+>(
+  options: CliOptionsT,
+  userOptions: GeneratorOptionsT
+): ParsedOptions<GeneratorOptionsT, CliOptionsT> {
+  return new OptionsParser(options, userOptions).parseOptionsWithDefaults();
+}
+
+class OptionsParser<
```

Comment:
[pp] There are two arguments maximum: CLI and user options. There you could create even a type if they are alwyas both needed. I think you could also pass the two around and use functions instead of the class. You moved the `getOptionsWithoutDefaults` already out of the class for testing.

Reply:
I changed a lot of things over time, I'll try this out 😊